File: leet_lint_code/leet_code/two_sum.cpp
// Copyright © 2018 Hailin Gu. All rights reserved.
// License(GPL)
// Author: Hailin Gu
// This is a answer of leet code problem 1.

#ifndef LEETLINTCODE_LINTCODE_TWOSUM_CPP
#define LEETLINTCODE_LINTCODE_TWOSUM_CPP

#include <stdio.h>
#include <vector>
#include <map>

class TwoSum {
 public:
   std::vector<int> twoSum(std::vector<int>& nums, int target) {
     std::vector<int> result;
     std::map<int, int> hashmap;

     size_t len = nums.size();
     for (int i = 0; i < len; ++ i) {
       if (hashmap.find(nums[i]) != hashmap.end()){
         int k = hashmap[nums[i]];
         result.push_back(i);
         if (k < i) result.insert(result.begin(), k);
         else result.push_back(k);
       } else hashmap[target - nums[i]] = i;
      }
      return result;
    }
};

#endif // LEETLINTCODE_LINTCODE_APLUSB_CPP

File: leet_lint_code/leet_code/java/TwoSum.java
// Copyright © 2018 Hailin Gu. All rights reserved.
// License(GPL)
// Author: Hailin Gu
// This is a answer of leet code problem 1.
// Date: 2019.7.2

import java.util.HashMap;
import java.util.Map;

class TwoSum {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        int[] ans = new int[2];

        for (int i = 0; i < nums.length; ++i) {
            if (map.containsKey(nums[i])) {
                ans[0] = map.get(nums[i]);
                ans[1] = i;
                break;
            } else {
                map.put(target - nums[i], i);
            }
        }
        return ans;
    }
}